src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs

```diff
@@ -4455,11 +4455,12 @@ internal SafeContext GetValEscape(BoundExpression expr, SafeContext scopeOfTheCo
 
                 case BoundKind.CollectionInitializerExpression:
                     var colExpr = (BoundCollectionInitializerExpression)expr;
-                    return GetValEscape(colExpr.Initializers, scopeOfTheContainingExpression);
-
-                case BoundKind.CollectionElementInitializer:
-                    var colElement = (BoundCollectionElementInitializer)expr;
-                    return GetValEscape(colElement.Arguments, scopeOfTheContainingExpression);
+                    // If arg mixing fails when the receiver has calling-method scope (i.e., some arguments could escape into the receiver), make the value scoped.
+                    return
+                        !scopeOfTheContainingExpression.IsConvertibleTo(SafeContext.CallingMethod) &&
```

Comment:
Why should we predicate this only on convertibility to `CallingMethod` scope? Consider the following:

```cs
RS M(ref int param)
{
    int local = 1;
    return new RS(ref param) { local }; // error expected
}

// ref struct RS, Add([UnscopedRef] in int item), etc.
```

In this case, `scopeOfTheContainingExpression` is `ReturnOnly` (right?), and not convertible to `CallingMethod`. But don't we still need to do a mixing check here?

Reply:
I could also imagine scenarios where `new RS(...)` has some local lifetime, but a collection initializer argument has an even narrower lifetime.

Reply:
The check `!scopeOfTheContainingExpression.IsConvertibleTo(SafeContext.CallingMethod)` is just an optimization - if it's false, it actually means `scopeOfTheContainingExpression == SafeContext.`, and we can return `SafeContext.CallingMethod` immediately.

I will refactor the code to make that clearer.

> In this case, `scopeOfTheContainingExpression` is `ReturnOnly` (right?), and not convertible to `CallingMethod`. But don't we still need to do a mixing check here?

We do a mixing check there. The `!scopeOfTheContainingExpression.IsConvertibleTo(SafeContext.CallingMethod)` condition is `true` in that case.

However, I'm not sure how that example is different from currently tested scenarios - you added a `ref param` to the ref struct's constructor, but that already goes through different code paths, in this PR I've only changed handling of the collection initializer, i.e., the part after the constructor - ` { local }`.

> I could also imagine scenarios where `new RS(...)` has some local lifetime, but a collection initializer argument has an even narrower lifetime.

I don't follow. Here we are computing ValEscape for the expression `new R() { ... }`, so it does not yet have any lifetime. And we either determine the expression to have ValEscape of `scopeOfTheContainingExpression` or `CallingMethod`. I don't think the collection initializer argument can have narrower lifetime than `scopeOfTheContainingExpression` (that's the narrowest lifetime we are currently at).